File: sdk/go/epoch/regional_bus.go
package epoch

import (
	"context"
	"fmt"
	"strings"
	"time"
)

const (
	maxRegionalBusDeliveryBatch = 100
	maxRegionalBusReadResults   = 10_000
	maxRegionalBusLongPoll      = 30 * time.Second
)

// SchemaFormat selects one compiler-backed schema language.
type SchemaFormat string

const (
	AvroSchema     SchemaFormat = "avro"
	JSONSchema     SchemaFormat = "json_schema"
	ProtobufSchema SchemaFormat = "protobuf"
)

// SchemaCompatibility controls adjacent-revision admission.
type SchemaCompatibility string

const (
	NoSchemaCompatibility       SchemaCompatibility = "none"
	BackwardSchemaCompatibility SchemaCompatibility = "backward"
	ForwardSchemaCompatibility  SchemaCompatibility = "forward"
	FullSchemaCompatibility     SchemaCompatibility = "full"
)

// SchemaRegistration is one compiler-validated schema revision request.
type SchemaRegistration struct {
	Name          string              `json:"name"`
	Format        SchemaFormat        `json:"format"`
	Definition    string              `json:"definition"`
	Compatibility SchemaCompatibility `json:"compatibility"`
	RootMessage   string              `json:"root_message,omitempty"`
}

// SchemaValidationMode selects producer advice, broker enforcement, both, or neither.
type SchemaValidationMode string

const (
	DisabledSchemaValidation          SchemaValidationMode = "disabled"
	ProducerSchemaValidation          SchemaValidationMode = "producer"
	BrokerSchemaValidation            SchemaValidationMode = "broker"
	ProducerAndBrokerSchemaValidation SchemaValidationMode = "producer_and_broker"
)

// SchemaValidationPolicy binds an event-type pattern to one immutable revision.
type SchemaValidationPolicy struct {
	Name             string               `json:"name"`
	EventTypePattern string               `json:"event_type_pattern"`
	SchemaRef        string               `json:"schema_ref"`
	Mode             SchemaValidationMode `json:"mode"`
}

// SchemaValidationStage selects an explicit read-only validation path.
type SchemaValidationStage string

const (
	ProducerValidationStage SchemaValidationStage = "producer"
	BrokerValidationStage   SchemaValidationStage = "broker"
)

// RegionalBusAcquireOptions identifies one bounded dispatcher lease request.
type RegionalBusAcquireOptions struct {
	Subscription    string
	Dispatcher      string
	DispatcherEpoch uint64
	MaxDeliveries   uint16
	Wait            time.Duration
}

// RegionalBusReplayOptions bounds one linearizable archive replay.
type RegionalBusReplayOptions struct {
	FromMS uint64
	ToMS   uint64
	Limit  uint16
	Filter *EventFilter
}

// RegionalBusDeliveryState filters the replicated delivery ledger.
type RegionalBusDeliveryState string

const (
	PendingDelivery      RegionalBusDeliveryState = "pending"
	InFlightDelivery     RegionalBusDeliveryState = "in_flight"
	AcknowledgedDelivery RegionalBusDeliveryState = "acknowledged"
	DeadLetteredDelivery RegionalBusDeliveryState = "dead_lettered"
)

// RegionalBusDeliveryQuery bounds one linearizable delivery-ledger query.
type RegionalBusDeliveryQuery struct {
	Subscription string
	State        *RegionalBusDeliveryState
	Limit        uint16
}

// RegionalBusClient routes authenticated Event Bus calls across regional nodes.
// Mutation retries preserve the caller's exact idempotency key.
type RegionalBusClient struct {
	regional *regionalClient
}

// NewRegionalBusClient builds a regional Event Bus client over HTTP endpoints.
func NewRegionalBusClient(endpoints []string, token string, scope RegionalScope, timeout time.Duration) (*RegionalBusClient, error) {
	regional, err := newRegionalClient(endpoints, token, scope, timeout)
	if err != nil {
		return nil, err
	}
	return &RegionalBusClient{regional: regional}, nil
}

// NewRegionalBusClientWithTransports injects transports for tests or custom networking.
func NewRegionalBusClientWithTransports(transports []Transport, token string, scope RegionalScope) (*RegionalBusClient, error) {
	regional, err := newRegionalClientWithTransports(transports, token, scope)
	if err != nil {
		return nil, err
	}
	return &RegionalBusClient{regional: regional}, nil
}

// UpsertSubscription atomically creates or replaces a typed route plan entry.
func (client *RegionalBusClient) UpsertSubscription(ctx context.Context, bus string, shard uint32, idempotencyKey string, subscription Subscription) (Document, error) {
	subscription, err := subscription.normalized()
	if err != nil {
		return nil, err
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind         string       `json:"kind"`
		Subscription Subscription `json:"subscription"`
	}{"upsert_subscription", subscription})
}

// RemoveSubscription deletes one exact route plan entry.
func (client *RegionalBusClient) RemoveSubscription(ctx context.Context, bus string, shard uint32, idempotencyKey, name string) (Document, error) {
	if strings.TrimSpace(name) == "" {
		return nil, fmt.Errorf("epoch: subscription name is required")
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind string `json:"kind"`
		Name string `json:"name"`
	}{"remove_subscription", name})
}

// Publish routes and archives one strict event envelope.
func (client *RegionalBusClient) Publish(ctx context.Context, bus string, shard uint32, idempotencyKey string, event EventEnvelope) (Document, error) {
	event, err := event.normalized()
	if err != nil {
		return nil, err
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind     string        `json:"kind"`
		Envelope EventEnvelope `json:"envelope"`
	}{"publish", event})
}

// AcquireDeliveries leases a bounded batch for one dispatcher epoch.
func (client *RegionalBusClient) AcquireDeliveries(ctx context.Context, bus string, shard uint32, idempotencyKey string, options RegionalBusAcquireOptions) (Document, error) {
	if strings.TrimSpace(options.Subscription) == "" {
		return nil, fmt.Errorf("epoch: subscription name is required")
	}
	if strings.TrimSpace(options.Dispatcher) == "" {
		return nil, fmt.Errorf("epoch: dispatcher is required")
	}
	if options.DispatcherEpoch == 0 {
		return nil, fmt.Errorf("epoch: dispatcher epoch must be non-zero")
	}
	if err := validateBusDeliveryBatch(options.MaxDeliveries); err != nil {
		return nil, err
	}
	if options.Wait < 0 || options.Wait > maxRegionalBusLongPoll || options.Wait%time.Millisecond != 0 {
		return nil, fmt.Errorf("epoch: delivery wait must be a whole number of milliseconds between 0 and %d", maxRegionalBusLongPoll.Milliseconds())
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind            string `json:"kind"`
		Subscription    string `json:"subscription"`
		Dispatcher      string `json:"dispatcher"`
		DispatcherEpoch string `json:"dispatcher_epoch"`
		MaxDeliveries   uint16 `json:"max_deliveries"`
		WaitMS          int64  `json:"wait_ms"`
	}{"acquire_deliveries", options.Subscription, options.Dispatcher, fmt.Sprintf("%d", options.DispatcherEpoch), options.MaxDeliveries, options.Wait.Milliseconds()})
}

// AcknowledgeDelivery permanently settles one fenced delivery lease.
func (client *RegionalBusClient) AcknowledgeDelivery(ctx context.Context, bus string, shard uint32, idempotencyKey, deliveryID, dispatcher string, dispatcherEpoch uint64, leaseToken string) (Document, error) {
	operation, err := busSettlement("acknowledge_delivery", deliveryID, dispatcher, dispatcherEpoch, leaseToken)
	if err != nil {
		return nil, err
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, operation)
}

// FailDelivery records a bounded failure and deterministic retry/dead-letter transition.
func (client *RegionalBusClient) FailDelivery(ctx context.Context, bus string, shard uint32, idempotencyKey, deliveryID, dispatcher string, dispatcherEpoch uint64, leaseToken, reason string) (Document, error) {
	operation, err := busSettlement("fail_delivery", deliveryID, dispatcher, dispatcherEpoch, leaseToken)
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(reason) == "" {
		return nil, fmt.Errorf("epoch: delivery failure reason is required")
	}
	operation["reason"] = reason
	return client.mutate(ctx, bus, shard, idempotencyKey, operation)
}

// RejectDelivery records a terminal failure and dead-letters the fenced attempt.
func (client *RegionalBusClient) RejectDelivery(ctx context.Context, bus string, shard uint32, idempotencyKey, deliveryID, dispatcher string, dispatcherEpoch uint64, leaseToken, reason string) (Document, error) {
	operation, err := busSettlement("reject_delivery", deliveryID, dispatcher, dispatcherEpoch, leaseToken)
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(reason) == "" {
		return nil, fmt.Errorf("epoch: delivery rejection reason is required")
	}
	operation["reason"] = reason
	return client.mutate(ctx, bus, shard, idempotencyKey, operation)
}

// RedriveDelivery returns one dead-lettered record to pending delivery with preserved history.
func (client *RegionalBusClient) RedriveDelivery(ctx context.Context, bus string, shard uint32, idempotencyKey, deliveryID string) (Document, error) {
	if strings.TrimSpace(deliveryID) == "" {
		return nil, fmt.Errorf("epoch: delivery ID is required")
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind       string `json:"kind"`
		DeliveryID string `json:"delivery_id"`
	}{"redrive_delivery", deliveryID})
}

// MaintainDeliveries applies due retry and expired-lease transitions explicitly.
func (client *RegionalBusClient) MaintainDeliveries(ctx context.Context, bus string, shard uint32, idempotencyKey string, maxDeliveries uint16) (Document, error) {
	if err := validateBusDeliveryBatch(maxDeliveries); err != nil {
		return nil, err
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind          string `json:"kind"`
		MaxDeliveries uint16 `json:"max_deliveries"`
	}{"maintain_deliveries", maxDeliveries})
}

// MaintainArchive applies bounded replicated age/count retention immediately.
func (client *RegionalBusClient) MaintainArchive(ctx context.Context, bus string, shard uint32, idempotencyKey string, maxEvents uint16) (Document, error) {
	if err := validateBusReadLimit(maxEvents); err != nil {
		return nil, err
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind      string `json:"kind"`
		MaxEvents uint16 `json:"max_events"`
	}{"maintain_archive", maxEvents})
}

// ApplyIntegration commits one typed schema, connector, MQTT, catalog, enrichment, or endpoint operation.
func (client *RegionalBusClient) ApplyIntegration(ctx context.Context, bus string, shard uint32, idempotencyKey string, operation Document) (Document, error) {
	if operation == nil {
		return nil, fmt.Errorf("epoch: integration operation is required")
	}
	kind, ok := operation["kind"].(string)
	if !ok || strings.TrimSpace(kind) == "" {
		return nil, fmt.Errorf("epoch: integration operation kind is required")
	}
	return client.mutate(ctx, bus, shard, idempotencyKey, struct {
		Kind      string   `json:"kind"`
		Operation Document `json:"operation"`
	}{"apply_integration", operation})
}

// RegisterSchema compiles and commits one immutable schema revision.
func (client *RegionalBusClient) RegisterSchema(ctx context.Context, bus string, shard uint32, idempotencyKey string, registration SchemaRegistration) (Document, error) {
	if err := registration.validate(); err != nil {
		return nil, err
	}
	return client.ApplyIntegration(ctx, bus, shard, idempotencyKey, Document{
		"kind":         "register_schema",
		"registration": registration,
	})
}

// UpsertSchemaValidationPolicy binds one event-type pattern to an immutable schema revision.
func (client *RegionalBusClient) UpsertSchemaValidationPolicy(ctx context.Context, bus string, shard uint32, idempotencyKey string, policy SchemaValidationPolicy) (Document, error) {
	if err := policy.validate(); err != nil {
		return nil, err
	}
	return client.ApplyIntegration(ctx, bus, shard, idempotencyKey, Document{
		"kind":   "upsert_validation_policy",
		"policy": policy,
	})
}

// RemoveSchemaValidationPolicy removes one exact validation binding.
func (client *RegionalBusClient) RemoveSchemaValidationPolicy(ctx context.Context, bus string, shard uint32, idempotencyKey, name string) (Document, error) {
	if !validResourceName(name) {
		return nil, fmt.Errorf("epoch: schema validation policy name is invalid")
	}
	return client.ApplyIntegration(ctx, bus, shard, idempotencyKey, Document{
		"kind": "remove_validation_policy",
		"name": name,
	})
}

// ValidateSchema performs a linearizable, read-only producer or broker validation.
// Broker-mode validation is also enforced atomically by Publish when configured.
func (client *RegionalBusClient) ValidateSchema(ctx context.Context, bus string, shard uint32, stage SchemaValidationStage, event EventEnvelope) (Document, error) {
	if stage != ProducerValidationStage && stage != BrokerValidationStage {
		return nil, fmt.Errorf("epoch: schema validation stage must be producer or broker")
	}
	event, err := event.normalized()
	if err != nil {
		return nil, err
	}
	return client.read(ctx, bus, shard, "POST", "/schema/validate", struct {
		Mode     SchemaValidationStage `json:"mode"`
		Envelope EventEnvelope         `json:"envelope"`
	}{stage, event})
}

// Mutation resolves one proposal from the current leader.
func (client *RegionalBusClient) Mutation(ctx context.Context, bus string, shard uint32, proposalID uint64) (Document, error) {
	if proposalID == 0 {
		return nil, fmt.Errorf("epoch: Event Bus proposal ID must be non-zero")
	}
	return client.read(ctx, bus, shard, "GET", fmt.Sprintf("/mutations/%d", proposalID), nil)
}

// ReplayArchive performs a bounded linearizable archive query.
func (client *RegionalBusClient) ReplayArchive(ctx context.Context, bus string, shard uint32, options RegionalBusReplayOptions) (Document, error) {
	if options.FromMS > options.ToMS {
		return nil, fmt.Errorf("epoch: Event Bus replay from time must not exceed to time")
	}
	if err := validateBusReadLimit(options.Limit); err != nil {
		return nil, err
	}
	body := map[string]any{"from_ms": fmt.Sprintf("%d", options.FromMS), "to_ms": fmt.Sprintf("%d", options.ToMS), "limit": options.Limit}
	if options.Filter != nil {
		body["filter"] = normalizedEventFilter(*options.Filter)
	}
	return client.read(ctx, bus, shard, "POST", "/archive/replay", body)
}

// QueryDeliveries returns bounded replicated delivery records.
func (client *RegionalBusClient) QueryDeliveries(ctx context.Context, bus string, shard uint32, query RegionalBusDeliveryQuery) (Document, error) {
	if err := validateBusReadLimit(query.Limit); err != nil {
		return nil, err
	}
	body := map[string]any{"limit": query.Limit}
	if strings.TrimSpace(query.Subscription) != "" {
		body["subscription"] = query.Subscription
	}
	if query.State != nil {
		if !query.State.valid() {
			return nil, fmt.Errorf("epoch: unsupported Event Bus delivery state %q", *query.State)
		}
		body["state"] = *query.State
	}
	return client.read(ctx, bus, shard, "POST", "/deliveries/query", body)
}

// Status returns the linearizable Event Bus tablet status and digest.
func (client *RegionalBusClient) Status(ctx context.Context, bus string, shard uint32) (Document, error) {
	return client.read(ctx, bus, shard, "GET", "/status", nil)
}

// IntegrationState returns the complete linearizable schema, connector, MQTT, catalog, and endpoint state.
func (client *RegionalBusClient) IntegrationState(ctx context.Context, bus string, shard uint32) (Document, error) {
	return client.read(ctx, bus, shard, "GET", "/integration/state", nil)
}

func (client *RegionalBusClient) mutate(ctx context.Context, bus string, shard uint32, idempotencyKey string, operation any) (Document, error) {
	if strings.TrimSpace(idempotencyKey) == "" {
		return nil, fmt.Errorf("epoch: idempotency key is required")
	}
	return regionalCall[Document](ctx, client.regionalClient(), "buses", "Event Bus", bus, shard, func(route regionalRoute) Request {
		return Request{Method: "POST", Path: "/mutations", Body: struct {
			IdempotencyKey string `json:"idempotency_key"`
			ExpectedTerm   string `json:"expected_term"`
			Operation      any    `json:"operation"`
		}{idempotencyKey, route.Term, operation}}
	})
}

func (client *RegionalBusClient) read(ctx context.Context, bus string, shard uint32, method, path string, body any) (Document, error) {
	return regionalCall[Document](ctx, client.regionalClient(), "buses", "Event Bus", bus, shard, func(_ regionalRoute) Request {
		return Request{Method: method, Path: path, Body: body, Headers: map[string]string{regionalReadHeader: "linearizable"}}
	})
}

func (client *RegionalBusClient) regionalClient() *regionalClient {
	if client == nil {
		return nil
	}
	return client.regional
}

func busSettlement(kind, deliveryID, dispatcher string, dispatcherEpoch uint64, leaseToken string) (map[string]any, error) {
	for label, value := range map[string]string{"delivery ID": deliveryID, "dispatcher": dispatcher, "delivery lease token": leaseToken} {
		if strings.TrimSpace(value) == "" {
			return nil, fmt.Errorf("epoch: %s is required", label)
		}
	}
	if dispatcherEpoch == 0 {
		return nil, fmt.Errorf("epoch: dispatcher epoch must be non-zero")
	}
	return map[string]any{"kind": kind, "delivery_id": deliveryID, "dispatcher": dispatcher, "dispatcher_epoch": fmt.Sprintf("%d", dispatcherEpoch), "lease_token": leaseToken}, nil
}

func validateBusDeliveryBatch(value uint16) error {
	if value == 0 || value > maxRegionalBusDeliveryBatch {
		return fmt.Errorf("epoch: Event Bus max deliveries must be between 1 and %d", maxRegionalBusDeliveryBatch)
	}
	return nil
}

func validateBusReadLimit(value uint16) error {
	if value == 0 || value > maxRegionalBusReadResults {
		return fmt.Errorf("epoch: Event Bus read limit must be between 1 and %d", maxRegionalBusReadResults)
	}
	return nil
}

func (state RegionalBusDeliveryState) valid() bool {
	return state == PendingDelivery || state == InFlightDelivery || state == AcknowledgedDelivery || state == DeadLetteredDelivery
}

func (registration SchemaRegistration) validate() error {
	if !validResourceName(registration.Name) {
		return fmt.Errorf("epoch: schema name is invalid")
	}
	if strings.TrimSpace(registration.Definition) == "" {
		return fmt.Errorf("epoch: schema definition is required")
	}
	if registration.Format != AvroSchema && registration.Format != JSONSchema && registration.Format != ProtobufSchema {
		return fmt.Errorf("epoch: unsupported schema format %q", registration.Format)
	}
	if registration.Compatibility != NoSchemaCompatibility && registration.Compatibility != BackwardSchemaCompatibility && registration.Compatibility != ForwardSchemaCompatibility && registration.Compatibility != FullSchemaCompatibility {
		return fmt.Errorf("epoch: unsupported schema compatibility %q", registration.Compatibility)
	}
	if registration.Format != ProtobufSchema && strings.TrimSpace(registration.RootMessage) != "" {
		return fmt.Errorf("epoch: root message is valid only for Protobuf schemas")
	}
	return nil
}

func (policy SchemaValidationPolicy) validate() error {
	if !validResourceName(policy.Name) {
		return fmt.Errorf("epoch: schema validation policy name is invalid")
	}
	if strings.TrimSpace(policy.EventTypePattern) == "" {
		return fmt.Errorf("epoch: schema validation event type pattern is required")
	}
	if strings.TrimSpace(policy.SchemaRef) == "" {
		return fmt.Errorf("epoch: schema reference is required")
	}
	if policy.Mode != DisabledSchemaValidation && policy.Mode != ProducerSchemaValidation && policy.Mode != BrokerSchemaValidation && policy.Mode != ProducerAndBrokerSchemaValidation {
		return fmt.Errorf("epoch: unsupported schema validation mode %q", policy.Mode)
	}
	return nil
}

func normalizedEventFilter(filter EventFilter) EventFilter {
	if filter.EventTypePatterns == nil {
		filter.EventTypePatterns = []string{}
	}
	if filter.SourcePatterns == nil {
		filter.SourcePatterns = []string{}
	}
	if filter.SubjectPatterns == nil {
		filter.SubjectPatterns = []string{}
	}
	if filter.Headers == nil {
		filter.Headers = map[string]string{}
	}
	if filter.JSONEquals == nil {
		filter.JSONEquals = map[string]any{}
	}
	return filter
}

File: sdk/python/src/epoch_sdk/regional_bus.py
"""Authenticated leader- and fence-aware regional Event Bus client."""

from __future__ import annotations

from typing import Any, Literal

from ._regional import RegionalClient, Route, _non_negative, _positive, _required
from .models import EventEnvelope, EventFilter, Subscription

DeliveryState = Literal["pending", "in_flight", "acknowledged", "dead_lettered"]
_DELIVERY_STATES = frozenset({"pending", "in_flight", "acknowledged", "dead_lettered"})
_MAX_DELIVERY_BATCH = 100
_MAX_READ_RESULTS = 10_000
_MAX_LONG_POLL_MS = 30_000
_LINEARIZABLE = {"x-epoch-read-consistency": "linearizable"}


class RegionalBusClient(RegionalClient):
    """Complete replicated Event Bus route, archive, and delivery-ledger client."""

    def upsert_subscription(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        subscription: Subscription,
    ) -> dict[str, Any]:
        if not isinstance(subscription, Subscription):
            raise TypeError("subscription must be Subscription")
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {
                "kind": "upsert_subscription",
                "subscription": subscription.to_dict(),
            },
        )

    def remove_subscription(
        self, bus: str, shard: int, idempotency_key: str, name: str
    ) -> dict[str, Any]:
        _required(name, "subscription name")
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {"kind": "remove_subscription", "name": name},
        )

    def publish(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        event: EventEnvelope,
    ) -> dict[str, Any]:
        if not isinstance(event, EventEnvelope):
            raise TypeError("event must be EventEnvelope")
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {"kind": "publish", "envelope": event.to_dict()},
        )

    def acquire_deliveries(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        *,
        subscription: str,
        dispatcher: str,
        dispatcher_epoch: int,
        max_deliveries: int,
        wait_ms: int = 0,
    ) -> dict[str, Any]:
        _required(subscription, "subscription name")
        _required(dispatcher, "dispatcher")
        _positive(dispatcher_epoch, "dispatcher epoch")
        _delivery_batch(max_deliveries)
        _non_negative(wait_ms, "delivery wait")
        if wait_ms > _MAX_LONG_POLL_MS:
            raise ValueError(f"delivery wait must not exceed {_MAX_LONG_POLL_MS} milliseconds")
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {
                "kind": "acquire_deliveries",
                "subscription": subscription,
                "dispatcher": dispatcher,
                "dispatcher_epoch": str(dispatcher_epoch),
                "max_deliveries": max_deliveries,
                "wait_ms": wait_ms,
            },
        )

    def acknowledge_delivery(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        delivery_id: str,
        dispatcher: str,
        dispatcher_epoch: int,
        lease_token: str,
    ) -> dict[str, Any]:
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            _settlement(
                "acknowledge_delivery",
                delivery_id,
                dispatcher,
                dispatcher_epoch,
                lease_token,
            ),
        )

    def fail_delivery(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        delivery_id: str,
        dispatcher: str,
        dispatcher_epoch: int,
        lease_token: str,
        reason: str,
    ) -> dict[str, Any]:
        _required(reason, "delivery failure reason")
        operation = _settlement(
            "fail_delivery", delivery_id, dispatcher, dispatcher_epoch, lease_token
        )
        operation["reason"] = reason
        return self._mutate(bus, shard, idempotency_key, operation)

    def reject_delivery(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        delivery_id: str,
        dispatcher: str,
        dispatcher_epoch: int,
        lease_token: str,
        reason: str,
    ) -> dict[str, Any]:
        _required(reason, "delivery rejection reason")
        operation = _settlement(
            "reject_delivery", delivery_id, dispatcher, dispatcher_epoch, lease_token
        )
        operation["reason"] = reason
        return self._mutate(bus, shard, idempotency_key, operation)

    def redrive_delivery(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        delivery_id: str,
    ) -> dict[str, Any]:
        """Return one dead-lettered delivery to pending with preserved history."""
        _required(delivery_id, "delivery ID")
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {"kind": "redrive_delivery", "delivery_id": delivery_id},
        )

    def maintain_deliveries(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        *,
        max_deliveries: int = _MAX_DELIVERY_BATCH,
    ) -> dict[str, Any]:
        _delivery_batch(max_deliveries)
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {"kind": "maintain_deliveries", "max_deliveries": max_deliveries},
        )

    def maintain_archive(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        *,
        max_events: int = _MAX_READ_RESULTS,
    ) -> dict[str, Any]:
        """Apply bounded replicated archive age/count retention immediately."""
        _read_limit(max_events)
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {"kind": "maintain_archive", "max_events": max_events},
        )

    def apply_integration(
        self,
        bus: str,
        shard: int,
        idempotency_key: str,
        operation: dict[str, Any],
    ) -> dict[str, Any]:
        """Commit one schema, connector, MQTT, catalog, enrichment, or endpoint operation."""
        if not isinstance(operation, dict) or not isinstance(operation.get("kind"), str):
            raise ValueError("integration operation kind is required")
        _required(operation["kind"], "integration operation kind")
        return self._mutate(
            bus,
            shard,
            idempotency_key,
            {"kind": "apply_integration", "operation": dict(operation)},
        )

    def mutation(self, bus: str, shard: int, proposal_id: int) -> dict[str, Any]:
        _positive(proposal_id, "Event Bus proposal ID")
        return self._read(bus, shard, "GET", f"/mutations/{proposal_id}")

    def replay_archive(
        self,
        bus: str,
        shard: int,
        *,
        from_ms: int,
        to_ms: int,
        limit: int = 100,
        filter: EventFilter | None = None,
    ) -> dict[str, Any]:
        _non_negative(from_ms, "Event Bus replay from time")
        _non_negative(to_ms, "Event Bus replay to time")
        if from_ms > to_ms:
            raise ValueError("Event Bus replay from time must not exceed to time")
        _read_limit(limit)
        body: dict[str, Any] = {
            "from_ms": str(from_ms),
            "to_ms": str(to_ms),
            "limit": limit,
        }
        if filter is not None:
            if not isinstance(filter, EventFilter):
                raise TypeError("Event Bus replay filter must be EventFilter")
            body["filter"] = filter.to_dict()
        return self._read(bus, shard, "POST", "/archive/replay", body)

    def query_deliveries(
        self,
        bus: str,
        shard: int,
        *,
        subscription: str | None = None,
        state: DeliveryState | None = None,
        limit: int = 100,
    ) -> dict[str, Any]:
        _read_limit(limit)
        body: dict[str, Any] = {"limit": limit}
        if subscription is not None:
            _required(subscription, "subscription name")
            body["subscription"] = subscription
        if state is not None:
            if state not in _DELIVERY_STATES:
                raise ValueError(f"unsupported Event Bus delivery state: {state}")
            body["state"] = state
        return self._read(bus, shard, "POST", "/deliveries/query", body)

    def status(self, bus: str, shard: int) -> dict[str, Any]:
        return self._read(bus, shard, "GET", "/status")

    def integration_state(self, bus: str, shard: int) -> dict[str, Any]:
        """Return the complete linearizable Event Bus integration state."""
        return self._read(bus, shard, "GET", "/integration/state")

    def _mutate(
        self, bus: str, shard: int, idempotency_key: str, operation: dict[str, Any]
    ) -> dict[str, Any]:
        _required(idempotency_key, "idempotency key")
        return self.call(
            "buses",
            "Event Bus",
            bus,
            shard,
            lambda route: _mutation_request(route, idempotency_key, operation),
        )

    def _read(
        self,
        bus: str,
        shard: int,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return self.call(
            "buses",
            "Event Bus",
            bus,
            shard,
            lambda _route: (method, path, body, None, _LINEARIZABLE),
        )


def _mutation_request(
    route: Route, idempotency_key: str, operation: dict[str, Any]
) -> tuple[str, str, dict[str, Any], None, dict[str, str]]:
    return (
        "POST",
        "/mutations",
        {
            "idempotency_key": idempotency_key,
            "expected_term": route.term,
            "operation": operation,
        },
        None,
        {},
    )


def _settlement(
    kind: str,
    delivery_id: str,
    dispatcher: str,
    dispatcher_epoch: int,
    lease_token: str,
) -> dict[str, Any]:
    _required(delivery_id, "delivery ID")
    _required(dispatcher, "dispatcher")
    _positive(dispatcher_epoch, "dispatcher epoch")
    _required(lease_token, "delivery lease token")
    return {
        "kind": kind,
        "delivery_id": delivery_id,
        "dispatcher": dispatcher,
        "dispatcher_epoch": str(dispatcher_epoch),
        "lease_token": lease_token,
    }


def _delivery_batch(value: int) -> None:
    if isinstance(value, bool) or not isinstance(value, int) or not 1 <= value <= 100:
        raise ValueError("Event Bus max deliveries must be between 1 and 100")


def _read_limit(value: int) -> None:
    if isinstance(value, bool) or not isinstance(value, int) or not 1 <= value <= 10_000:
        raise ValueError("Event Bus read limit must be between 1 and 10000")


__all__ = ["DeliveryState", "RegionalBusClient"]
